CMake/ExternalData.cmake

```diff
@@ -0,0 +1,1202 @@
+
+  The ``ExternalData_Expand_Arguments`` function evaluates ``DATA{}``
+  references in its arguments and constructs a new list of arguments::
+
+    ExternalData_Expand_Arguments(
+      <target>   # Name of data management target
+      <outVar>   # Output variable
+      [args...]  # Input arguments, DATA{} allowed
+      )
+
+  It replaces each ``DATA{}`` reference in an argument with the full path of
+  a real data file on disk that will exist after the ``<target>`` builds.
+
+.. command:: ExternalData_Add_Test
+
+  The ``ExternalData_Add_Test`` function wraps around the CMake
+  :command:`add_test` command but supports ``DATA{}`` references in
+  its arguments::
+
+    ExternalData_Add_Test(
+      <target>   # Name of data management target
+      ...        # Arguments of add_test(), DATA{} allowed
+      )
+
+  It passes its arguments through ``ExternalData_Expand_Arguments`` and then
+  invokes the :command:`add_test` command using the results.
+
+.. command:: ExternalData_Add_Target
+
+  The ``ExternalData_Add_Target`` function creates a custom target to
+  manage local instances of data files stored externally::
+
+    ExternalData_Add_Target(
+      <target>   # Name of data management target
+      )
+
+  It creates custom commands in the target as necessary to make data
+  files available for each ``DATA{}`` reference previously evaluated by
+  other functions provided by this module.
+  Data files may be fetched from one of the URL templates specified in
+  the ``ExternalData_URL_TEMPLATES`` variable, or may be found locally
+  in one of the paths specified in the ``ExternalData_OBJECT_STORES``
+  variable.
+
+  Typically only one target is needed to manage all external data within
+  a project.  Call this function once at the end of configuration after
+  all data references have been processed.
+
+Module Variables
+^^^^^^^^^^^^^^^^
+
+The following variables configure behavior.  They should be set before
+calling any of the functions provided by this module.
+
+.. variable:: ExternalData_BINARY_ROOT
+
+  The ``ExternalData_BINARY_ROOT`` variable may be set to the directory to
+  hold the real data files named by expanded ``DATA{}`` references.  The
+  default is ``CMAKE_BINARY_DIR``.  The directory layout will mirror that of
+  content links under ``ExternalData_SOURCE_ROOT``.
+
+.. variable:: ExternalData_CUSTOM_SCRIPT_<key>
+
+  Specify a full path to a ``.cmake`` custom fetch script identified by
+  ``<key>`` in entries of the ``ExternalData_URL_TEMPLATES`` list.
+  See `Custom Fetch Scripts`_.
+
+.. variable:: ExternalData_LINK_CONTENT
+
+  The ``ExternalData_LINK_CONTENT`` variable may be set to the name of a
+  supported hash algorithm to enable automatic conversion of real data
+  files referenced by the ``DATA{}`` syntax into content links.  For each
+  such ``<file>`` a content link named ``<file><ext>`` is created.  The
+  original file is renamed to the form ``.ExternalData_<algo>_<hash>`` to
```

Comment:
List of algos:
[`73e6796` (#3322)](https://github.com/InsightSoftwareConsortium/ITK/pull/3322/commits/73e67962a11bdbe2bed75590130631a2f6d60387#diff-ad081f045f7b916dc1419f5571ae3b4fa3a2612ae97135a4f27cf3c4b3829e0cR276-R285)

Reply:
So a filename could look something like `.ExternalData_MD5_some_long_hash`? I don't know that I fully understand how content links are created or used.

Reply:
As I understood, pretty much everything is same as before, from user's perspective. There is now support for IPFS via .cid extension. Some of the machinery might be new and more flexible, or that might all be the same as before, as I never used anything but the [simplest few cases](https://github.com/InsightSoftwareConsortium/ITKMontage/blob/v0.7.3/test/CMakeLists.txt#L173-L182).

Reply:
Files are created, named per hashes. For example, on my computer there is a file called `C:\Temp\ITKExternalData\SHA512\0b6dad04a7d2d54c56899a4c620a89f7c83d1713ac55cb4f8327f8a81c93e506b2cf2761c60e54b10c0221b620cc458b6901c158c9070480cfe5dc7d601ee6cb`. I have environment variable called `ExternalData_OBJECT_STORES` pointing to `C:/Temp/ITKExternalData`, in order to avoid having this cache in every ITK build tree.

Reply:
Ah okay, I see those objects in my `ITK-build/ExternalData/Objects/SHA512` folder as well. Thanks.